File: app/lib/device/tempRefreshAudit.ts
// TEMP_REFRESH_AUDIT: isolated temporary validation/classification used by the
// authenticated physical-device ingestion route. Remove with the audit system.
type TEMP_REFRESH_AUDIT_UnknownRecord = Record<string, unknown>

export type TEMP_REFRESH_AUDIT_Record = TEMP_REFRESH_AUDIT_UnknownRecord & { event_seq: number }

export const TEMP_REFRESH_AUDIT_ENABLED = process.env.TEMP_REFRESH_AUDIT_ENABLED === 'true'

export const TEMP_REFRESH_AUDIT_DECISIONS = new Set([
  'filtered_change', 'useful_redraw', 'wasted_redraw', 'no_redraw', 'avoidable_wake',
  'intentional_refresh', 'display_failed',
])
const TEMP_REFRESH_AUDIT_MIN_VALID_UNIX_TIME = 1577836800 // 2020-01-01 UTC
const TEMP_REFRESH_AUDIT_MAX_VALID_UNIX_TIME = 4102444800 // 2100-01-01 UTC

export function TEMP_REFRESH_AUDIT_classify(record: TEMP_REFRESH_AUDIT_UnknownRecord): string {
  if (record.display_attempted === true && record.display_succeeded === false) return 'display_failed'
  if ((record.metadata as TEMP_REFRESH_AUDIT_UnknownRecord | undefined)?.intentional_refresh === true) return 'intentional_refresh'
  if (record.physical_refresh === true) return record.render_changed === false ? 'wasted_redraw' : 'useful_redraw'
  if (record.render_changed === false && record.backend_revision_before !== record.backend_revision_after) return 'filtered_change'
  const metadata = record.metadata as TEMP_REFRESH_AUDIT_UnknownRecord | undefined
  const confidentlyAvoidableTimerWake = metadata?.avoidable_wake_confident === true &&
    record.trigger === 'scheduled_revision_poll' &&
    record.wake_reason === 'timer' && typeof record.module === 'string' && record.module.length > 0 &&
    record.backend_revision_before === record.backend_revision_after && record.render_changed === false &&
    record.display_attempted === false
  if (confidentlyAvoidableTimerWake) return 'avoidable_wake'
  return 'no_redraw'
}

export function TEMP_REFRESH_AUDIT_sanitize(record: unknown): TEMP_REFRESH_AUDIT_Record | null {
  if (!record || typeof record !== 'object' || Array.isArray(record)) return null
  const input = record as TEMP_REFRESH_AUDIT_UnknownRecord
  const text = (key: string, max = 512): string | null => typeof input[key] === 'string' ? String(input[key]).slice(0, max) : null
  const jsonObject = (key: string, fallback: object): object =>
    input[key] !== null && typeof input[key] === 'object' ? input[key] as object : fallback
  const refreshType = text('refresh_type', 16)
  if (!['none', 'partial', 'full'].includes(refreshType ?? '')) return null
  if (typeof input.display_attempted !== 'boolean') return null
  const displayAttempted = input.display_attempted
  const displaySucceeded = typeof input.display_succeeded === 'boolean' ? input.display_succeeded : null
  const attemptedType = String(input.refresh_type_attempted)
  const physicalRefresh = input.physical_refresh === true
  if (displayAttempted) {
    if (displaySucceeded === null || !['partial', 'full'].includes(attemptedType)) return null
  } else if (displaySucceeded !== null || attemptedType !== 'none') return null
  if (physicalRefresh !== (displayAttempted && displaySucceeded === true)) return null
  if ((physicalRefresh && refreshType !== attemptedType) || (!physicalRefresh && refreshType !== 'none')) return null
  const eventSeq = input.event_seq
  if (typeof eventSeq !== 'number' || !Number.isSafeInteger(eventSeq) || eventSeq < 1) return null
  let occurredAt: string | null = null
  if (input.occurred_at != null) {
    // TEMP_REFRESH_AUDIT: a corrupt RTC must not reject otherwise useful facts.
    // Match firmware bounds and store an invalid device event time as null.
    if (typeof input.occurred_at === 'number' && Number.isSafeInteger(input.occurred_at) &&
        input.occurred_at >= TEMP_REFRESH_AUDIT_MIN_VALID_UNIX_TIME &&
        input.occurred_at <= TEMP_REFRESH_AUDIT_MAX_VALID_UNIX_TIME) {
      occurredAt = new Date(input.occurred_at * 1000).toISOString()
    }
  }
  const sanitized: TEMP_REFRESH_AUDIT_Record = {
    event_seq: eventSeq, occurred_at: occurredAt,
    firmware_version: text('firmware_version', 64), trigger: text('trigger', 64) ?? 'other',
    source: text('source', 128), module: text('module', 128),
    raw_changes: jsonObject('raw_changes', {}), display_changes: jsonObject('display_changes', {}),
    previous_render_hash: text('previous_render_hash', 128), new_render_hash: text('new_render_hash', 128),
    render_changed: typeof input.render_changed === 'boolean' ? input.render_changed : null,
    physical_refresh: physicalRefresh,
    display_attempted: displayAttempted,
    display_succeeded: displaySucceeded,
    refresh_type_attempted: attemptedType,
    refresh_type: refreshType,
    dirty_regions: Array.isArray(input.dirty_regions) ? input.dirty_regions.slice(0, 16) : [],
    decision_reason: text('decision_reason'),
    backend_revision_before: Number.isSafeInteger(input.backend_revision_before) ? input.backend_revision_before : null,
    backend_revision_after: Number.isSafeInteger(input.backend_revision_after) ? input.backend_revision_after : null,
    battery_percent: typeof input.battery_percent === 'number' ? input.battery_percent : null,
    battery_voltage: typeof input.battery_voltage === 'number' ? input.battery_voltage : null,
    charger_connected: typeof input.charger_connected === 'boolean' ? input.charger_connected : null,
    wake_reason: text('wake_reason', 64), metadata: jsonObject('metadata', {}),
  }
  sanitized.decision = TEMP_REFRESH_AUDIT_classify(sanitized)
  return sanitized
}

// TEMP_REFRESH_AUDIT: behavioural batch boundary shared by route tests. The
// unique event sequence also makes duplicate records within one resend harmless.
export function TEMP_REFRESH_AUDIT_prepareBatch(records: unknown, enabled = TEMP_REFRESH_AUDIT_ENABLED): { error: string | null, records: TEMP_REFRESH_AUDIT_Record[] } {
  if (!enabled) return { error: 'disabled', records: [] }
  if (!Array.isArray(records) || records.length < 1 || records.length > 8) return { error: 'invalid_batch', records: [] }
  const sanitized = records.map(TEMP_REFRESH_AUDIT_sanitize)
  if (sanitized.some((record) => record === null)) return { error: 'invalid_record', records: [] }
  const validRecords = sanitized.filter((record): record is TEMP_REFRESH_AUDIT_Record => record !== null)
  const unique = [...new Map(validRecords.map((record) => [record.event_seq, record])).values()]
  return { error: null, records: unique }
}

File: frame/src/core/TempRefreshAudit.cpp
#include "TempRefreshAudit.h"
#include "Config.h"
#if TEMP_REFRESH_AUDIT_ENABLED
#include "DeviceIdentity.h"
#include "NetClient.h"
#include <ArduinoJson.h>
#include <Preferences.h>
#include <WiFi.h>

namespace {
constexpr uint8_t TEMP_REFRESH_AUDIT_MAX_RECORDS = 8;
constexpr uint8_t TEMP_REFRESH_AUDIT_BATTERY_BATCH_THRESHOLD = 4;
// TEMP_REFRESH_AUDIT: use an explicit 64-bit type. Arduino-ESP32 2.0.14 builds
// some targets with a 32-bit time_t, so 2100 cannot be a time_t constant.
constexpr int64_t TEMP_REFRESH_AUDIT_MIN_VALID_UNIX_TIME = 1577836800LL; // 2020-01-01 UTC
constexpr int64_t TEMP_REFRESH_AUDIT_MAX_VALID_UNIX_TIME = 4102444800LL; // 2100-01-01 UTC
Preferences auditPrefs;

void persistTEMP_REFRESH_AUDIT(JsonDocument& doc) {
  auditPrefs.begin("refresh_audit", false);
  const uint64_t eventSeq = auditPrefs.getULong64("event_seq", 0) + 1;
  auditPrefs.putULong64("event_seq", eventSeq); doc["event_seq"] = eventSeq;
  const int64_t now = static_cast<int64_t>(time(nullptr));
  if (now >= TEMP_REFRESH_AUDIT_MIN_VALID_UNIX_TIME &&
      now <= TEMP_REFRESH_AUDIT_MAX_VALID_UNIX_TIME) doc["occurred_at"] = (int64_t)now;
  else doc["occurred_at"] = nullptr;
  String encoded; serializeJson(doc, encoded);
  uint8_t head = auditPrefs.getUChar("head", 0), count = auditPrefs.getUChar("count", 0);
  char key[5]; snprintf(key, sizeof(key), "r%u", head); auditPrefs.putString(key, encoded);
  auditPrefs.putUChar("head", (head + 1) % TEMP_REFRESH_AUDIT_MAX_RECORDS);
  auditPrefs.putUChar("count", count < TEMP_REFRESH_AUDIT_MAX_RECORDS ? count + 1 : TEMP_REFRESH_AUDIT_MAX_RECORDS);
  auditPrefs.end();
}
}

void TempRefreshAudit::TEMP_REFRESH_AUDIT_record(const char* trigger, const String& sourceModules,
  const SmartRenderState& desired, const String& previousPhysicalHash,
  const SmartDisplayPlan& plan, bool displaySucceeded,
  uint64_t backendBefore, uint64_t backendAfter, const BatteryState& battery,
  bool chargerConnected, const char* wakeReason, const char* firmwareVersion) {
  // Persist compact facts only. Module hashes are the existing normalized render
  // model; no second renderer is introduced by TEMP_REFRESH_AUDIT.
  const String previous = previousPhysicalHash;
  const String next = SmartRefresh::TEMP_REFRESH_AUDIT_renderHash(desired);
  const bool changed = !previous.length() || previous != next;
  const bool physical = plan.type != SmartDisplayPlan::NONE && displaySucceeded;
  const bool displayAttempted = plan.type != SmartDisplayPlan::NONE;
  const bool sourceChanged = backendAfter != backendBefore;
  const bool avoidableWake = !displayAttempted && !changed && !sourceChanged &&
    strcmp(trigger, "scheduled_revision_poll") == 0 && sourceModules.length() > 0 &&
    strcmp(wakeReason, "timer") == 0;
  const char* decision = displayAttempted && !displaySucceeded ? "display_failed"
    : physical ? (changed ? "useful_redraw" : "wasted_redraw")
    : avoidableWake ? "avoidable_wake"
    : (!changed && sourceChanged ? "filtered_change" : "no_redraw");
  DynamicJsonDocument doc(2048);
  doc["firmware_version"] = firmwareVersion; doc["trigger"] = trigger;
  doc["source"] = sourceModules; doc["module"] = sourceModules;
  JsonObject display = doc.createNestedObject("display_changes");
  display["canonical_modules_before"] = previous; display["canonical_modules_after"] = next;
  doc.createNestedObject("raw_changes"); // Raw snapshots are nullable/unavailable on-device.
  doc["previous_render_hash"] = previous; doc["new_render_hash"] = next;
  doc["render_changed"] = changed; doc["physical_refresh"] = physical;
  doc["display_attempted"] = displayAttempted;
  if (displayAttempted) doc["display_succeeded"] = displaySucceeded;
  else doc["display_succeeded"] = nullptr;
  doc["refresh_type_attempted"] = displayAttempted ? (plan.type == SmartDisplayPlan::FULL ? "full" : "partial") : "none";
  doc["refresh_type"] = physical ? (plan.type == SmartDisplayPlan::FULL ? "full" : "partial") : "none";
  JsonArray regions = doc.createNestedArray("dirty_regions");
  for (uint8_t i = 0; i < plan.regionCount; ++i) { JsonObject r = regions.createNestedObject(); r["x"] = plan.regions[i].x; r["y"] = plan.regions[i].y; r["w"] = plan.regions[i].w; r["h"] = plan.regions[i].h; }
  doc["decision"] = decision;
  doc["decision_reason"] = displayAttempted && !displaySucceeded ? "Required physical display update was attempted but failed"
    : physical ? (changed ? "Physical refresh changed canonical rendered output" : "Physical refresh occurred despite unchanged canonical render hash")
    : avoidableWake ? "Timer woke for due module evaluation but source revision and canonical display output were unchanged"
    : (!changed && sourceChanged ? "Source changed but normalized render output was identical" : "Evaluation required no physical display update");
  doc["backend_revision_before"] = backendBefore; doc["backend_revision_after"] = backendAfter;
  doc["battery_percent"] = battery.percent; doc["battery_voltage"] = battery.smoothedVoltage;
  doc["charger_connected"] = chargerConnected; doc["wake_reason"] = wakeReason;
  persistTEMP_REFRESH_AUDIT(doc);
}

void TempRefreshAudit::TEMP_REFRESH_AUDIT_recordIntentionalRefresh(const char* trigger,
  const char* reason, bool displaySucceeded, const BatteryState& battery,
  bool chargerConnected, const char* wakeReason, const char* firmwareVersion) {
  DynamicJsonDocument doc(1024);
  doc["firmware_version"] = firmwareVersion; doc["trigger"] = trigger;
  doc["source"] = "local_display"; doc["module"] = "display";
  doc.createNestedObject("raw_changes"); doc.createNestedObject("display_changes");
  doc["previous_render_hash"] = nullptr; doc["new_render_hash"] = nullptr;
  doc["render_changed"] = nullptr; doc["physical_refresh"] = displaySucceeded;
  doc["display_attempted"] = true; doc["display_succeeded"] = displaySucceeded;
  doc["refresh_type_attempted"] = "full";
  doc["refresh_type"] = displaySucceeded ? "full" : "none";
  doc.createNestedArray("dirty_regions"); doc["decision"] = "intentional_refresh";
  doc["decision_reason"] = reason; doc["battery_percent"] = battery.percent;
  doc["battery_voltage"] = battery.smoothedVoltage; doc["charger_connected"] = chargerConnected;
  doc["wake_reason"] = wakeReason; JsonObject metadata = doc.createNestedObject("metadata");
  metadata["intentional_refresh"] = true;
  persistTEMP_REFRESH_AUDIT(doc);
}

void TempRefreshAudit::TEMP_REFRESH_AUDIT_flushPiggyback(const String& token, bool force) {
  if (WiFi.status() != WL_CONNECTED) return; // Never creates a network session.
  auditPrefs.begin("refresh_audit", false);
  uint8_t head = auditPrefs.getUChar("head", 0), count = auditPrefs.getUChar("count", 0);
  if (!count) { auditPrefs.end(); return; }
  if (!force && count < TEMP_REFRESH_AUDIT_BATTERY_BATCH_THRESHOLD) { auditPrefs.end(); return; }
  DynamicJsonDocument batch(16384); batch["device_id"] = DeviceIdentity::getDeviceId(); JsonArray records = batch.createNestedArray("records");
  const uint8_t start = (head + TEMP_REFRESH_AUDIT_MAX_RECORDS - count) % TEMP_REFRESH_AUDIT_MAX_RECORDS;
  for (uint8_t i = 0; i < count; ++i) {
    char key[5]; snprintf(key, sizeof(key), "r%u", (start + i) % TEMP_REFRESH_AUDIT_MAX_RECORDS);
    String raw = auditPrefs.getString(key, "");
    if (raw.length()) { DynamicJsonDocument item(2048); if (!deserializeJson(item, raw)) records.add(item.as<JsonObject>()); }
  }
  String body; serializeJson(batch, body); int code = 0; String response;
  const bool sent = NetClient::TEMP_REFRESH_AUDIT_httpPostConnected(String(BASE_URL) + "/api/device/refresh-audit", token, body, code, response);
  if (sent && code >= 200 && code < 300) { for (uint8_t i = 0; i < TEMP_REFRESH_AUDIT_MAX_RECORDS; ++i) { char key[5]; snprintf(key, sizeof(key), "r%u", i); auditPrefs.remove(key); } auditPrefs.putUChar("count", 0); }
  auditPrefs.end();
}
#endif // TEMP_REFRESH_AUDIT_ENABLED
